lib/vtls/schannel.c

```diff
@@ -708,22 +727,235 @@ schannel_acquire_credential_handle(struct Curl_easy *data,
     return CURLE_OUT_OF_MEMORY;
   }
   backend->cred->refcount = 1;
 
-  sspi_status =
-    s_pSecFn->AcquireCredentialsHandle(NULL, (TCHAR *)UNISP_NAME,
-                                       SECPKG_CRED_OUTBOUND, NULL,
-                                       &schannel_cred, NULL, NULL,
-                                       &backend->cred->cred_handle,
-                                       &backend->cred->time_stamp);
+  /* Windows 10, 1809 (a.k.a. Windows 10 build 17763) */
+  if(curlx_verify_windows_version(10, 0, 17763, PLATFORM_WINNT,
+      VERSION_GREATER_THAN_EQUAL)) {
+
+    int disable_aes_gcm_sha384 = 0;
+    int disable_aes_gcm_sha256 = 0;
+    int disable_chacha_poly = 1;
+
+    SCH_CREDENTIALS credentials = { 0 };
+    TLS_PARAMETERS tls_parameters = { 0 };
+    CRYPTO_SETTINGS crypto_settings[4] = { 0 };
+    UNICODE_STRING blocked_ccm_modes[1] = { 0 };
+    UNICODE_STRING blocked_gcm_modes[1] = { 0 };
+
+    /*
+     RTM Windows 2022 / Windows 11 does not support ChaCha20-Poly1305
+     or AES_CCM. So disable both those options explictly (TLS 1.3
+     includes those as 2 built-in Cipher-Suites).
+
+     Add conditional curlx_verify_windows_version() calls when
+     support has been added.
+    */
+
+    int crypto_settings_idx = 0;
 
+
+    /*
+      Disallow AES_CCM algorithm, since there's no support for it in SChannel.
+    */
+    blocked_ccm_modes[0] = (UNICODE_STRING) {
+        sizeof(BCRYPT_CHAIN_MODE_CCM),
+        sizeof(BCRYPT_CHAIN_MODE_CCM),
+        BCRYPT_CHAIN_MODE_CCM };
+
+    crypto_settings[crypto_settings_idx].eAlgorithmUsage =
+                                             TlsParametersCngAlgUsageCipher;
+    crypto_settings[crypto_settings_idx].rgstrChainingModes =
+                                             blocked_ccm_modes;
+    crypto_settings[crypto_settings_idx].cChainingModes =
+                                             ARRAYSIZE(blocked_ccm_modes);
+    crypto_settings[crypto_settings_idx].strCngAlgId = (UNICODE_STRING) {
+        sizeof(BCRYPT_AES_ALGORITHM),
+        sizeof(BCRYPT_AES_ALGORITHM),
+        BCRYPT_AES_ALGORITHM };
+    crypto_settings_idx++;
+
+
+    /* Only 3 ciphers for TLS 1.3 remain:
+    * TLS_AES_256_GCM_SHA384
+    * TLS_AES_128_GCM_SHA256
+    * TLS_CHACHA20_POLY1305_SHA256
+    */
+    char *ciphers13 = SSL_CONN_CONFIG(cipher_list13);
+    if(ciphers13) {
+      int remaining_ciphers = 3;
+
+      /* detect which remaining ciphers to enable
+      and then disable everything else.
+      */
+
+      char *startCur = ciphers13;
+      int algCount = 0;
+      char tmp[LONGEST_ALG_ID] = { 0 };
+      char *nameEnd;
+
+      disable_aes_gcm_sha384 = 1;
+      disable_aes_gcm_sha256 = 1;
+
+      while(startCur && (0 != *startCur) && (algCount < remaining_ciphers)) {
+
+        nameEnd = strchr(startCur, ':');
+        size_t n = nameEnd ? \
+            min((size_t)(nameEnd - startCur), LONGEST_ALG_ID - 1) : \
+            min(strlen(startCur), LONGEST_ALG_ID - 1);
```

Comment:
This is some hairy logic that is terribly hard to follow. But perhaps more importantly, someone provides a too long name here, you just truncate it and use the remainder as a cipher name. I think that's wrong. I think a too long name is an error!

Reply:
You're right. I'll change this and the similar logic used in the pre-TLS 1.3 cipher selection (in `get_alg_id_by_name()`).